### PR TITLE
Adding unique_placement_prefix to ensure matching tags are not used twice

### PR DESCRIPTION
**Pull request by matthiashanel**
Allows to not place a stream in the same availability zone twice

Signed-off-by: Matthias Hanel <mh@synadia.com>

### CODE REVIEW

**Comment by kozlovic** [comment]
@matthiashanel Could you describe a bit in the PR description how the feature would be used and give an example. I am looking at the test case added (which is good) but I don't have a sense of how this works. Having more details in the PR can then be used as base for documentation, so it is not wasted. Thanks!

**Comment by matthiashanel** [comment]
@kozlovic the new filed specifies a mandatory prefix of a server_tag used for placement.
When the server makes a placement decision in addition to other factors (such as placment tags provided, or disk utilization) it picks only server that have `server_tags` with the specified value as prefix and where the matching tag is unique across all selected server.  

Say a three server have the server tags below and the `unique_placement_prefix` is `az`:
```
server 1: server_tags: [az:1]
server 2: server_tags: [az:2]
server 3: server_tags: [az:3]
server 4: server_tags: [az:1]
```

An R3 stream would only be placed on servers 1-3 or 2-4. Never would server 1 and 4 be used for to host the same R3 stream.

**Comment by kozlovic** [comment]
> An R3 stream would only be placed on servers 1-3 or 2-4. Never would server 1 and 4 be used for to host the same R3 stream.

Thanks, that helps a lot. Will review...

**Review by kozlovic — APPROVED** [review]
LGTM

File: server/jetstream_cluster_test.go [hunk 1]
```diff
@@ -4563,6 +4563,85 @@ func TestJetStreamClusterSuperClusterMetaPlacement(t *testing.T) {
 	}
 }
 
+func TestJetStreamUniquePlacementTag(t *testing.T) {
+	tmlp := `
+		listen: 127.0.0.1:-1
+		server_name: %s
+		jetstream: {max_mem_store: 256MB, max_file_store: 2GB, store_dir: '%s', limits: {unique_placement_prefix: az}}
```
Comment (derekcollison) [inline server/jetstream_cluster_test.go:4570]: This feels like its not a descendant of limits imo. Just inline with max_file_store etc. But could be wrong.

File: server/opts.go [hunk 2]
```diff
@@ -178,8 +178,9 @@ type RemoteLeafOpts struct {
 }
 
 type JSLimitOpts struct {
-	MaxAckPending int
-	Duplicates    time.Duration
+	MaxAckPending            int
+	Duplicates               time.Duration
+	UniquePlacementTagPrefix string
```
Comment (derekcollison) [inline server/opts.go:183]: These are per server yes?

How about UniqueTags?

File: server/opts.go [hunk 3]
```diff
@@ -1813,6 +1814,8 @@ func parseJetStreamLimits(v interface{}, opts *Options, errors *[]error, warning
 			if err != nil {
 				*errors = append(*errors, err)
 			}
+		case "unique_placement_prefix":
```
Comment (derekcollison) [inline server/opts.go:1817]: unique_tags imo is good.

**Review by derekcollison — APPROVED** [review]
LGTM

**Comment by samuel-form3** [comment]
Awesome stuff @matthiashanel @derekcollison 

Thanks, we'll give you feedback afterwards

**Pull request merged**